torch/utils/data/datapipes/iter/combining.py

```diff
@@ -55,16 +57,97 @@ def __iter__(self):
 
 @functional_datapipe('fork')
 class ForkIterDataPipe(IterDataPipe):
+    r""" :class:`ForkIterDataPipe`.
+
+        Iterable DataPipe to create multiple instances of the same Iterable DataPipe.
+        args:
+            datapipe: Iterable DataPipe being copied
+            num_instances: number of instances of the datapipe to create
+            buffer_size: this restricts how far ahead the leading child DataPipe
+             can read relative to the slowest child DataPipe
+    """
+    def __new__(cls, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        container = _ForkIterDataPipe(datapipe, num_instances, buffer_size)
+        return [ChildDataPipe(container, i) for i in range(num_instances)]
 
-    def __new__(cls, datapipe, instances):
-        result = []
-        buffer = list(datapipe)
-        return [IterateBuffer(buffer) for i in range(instances)]
+    def __init__(self, *arg):
+        raise Exception("__init__ called instead of __new__")
+
+
+class _ForkIterDataPipe(IterDataPipe):
+    r""" :class:`_ForkIterDataPipe`.
 
+        Container to hold instance-specific information on behalf of ForkIterDataPipe. It tracks
+        the state of its child DataPipes, maintains the buffer, and yields the next value
+        as requested by the child DataPipes.
+    """
+    def __init__(self, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        self.main_datapipe = iter(datapipe)
```

Comment:
no need to call iter here, just use datapipe, you are invoking unnecessary operation here

Reply:
Without calling `iter`, how do I get the next element on line 107 (see below)?

`self.buffer.append(self.main_datapipe.__next__())`

Reply:
Sounds reasonable, we will need to take care of reiterating later.

Reply:
Actually, this can be the reason of pickle failure, you could try to do lazy initialization of iter(datapipe) inside of get_next

Reply:
Yep, it seems to be the case. I am getting different errors now.

Reply:
I am switching to a lazy initialization of `iter(datapipe)` within the method `get_next` to address the issue with pickle.